### PR TITLE
Add randint_like support (should be final randomness function! 🥳)

### PR DESCRIPTION
**Pull request by samdow**

### CODE REVIEW

File: test/test_vmap.py [hunk 1]
```diff
@@ -3590,7 +3575,7 @@ def test_feature_dropout(self, device, randomness):
             vmap_result = vmap(op, randomness=randomness)(passed)
             torch.manual_seed(seed)
             if randomness == 'different':
-                expected = op(passed.flatten(0, 1))
+                expected = op(passed.flatten(0, 1)) # (B0, B, ...) -> (B0 * B, ...)
```
Comment (zou3519) [inline test/test_vmap.py:3578]: flake8 alert (not sure if it's this line, just needed to put the comment somewhere)

File: functorch/csrc/BatchRulesRandomness.cpp [hunk 2]
```diff
@@ -149,6 +149,33 @@ Tensor unary_pointwise_random_batch_rule(const Tensor& tensor, ExtraArgs... extr
   return makeBatched(out, 0, cur_level);
 }
 
+template<typename... ExtraArgs>
+Tensor randint_like_random_batch_rule(const Tensor& self, ExtraArgs... extra_args) {
```
Comment (zou3519) [inline functorch/csrc/BatchRulesRandomness.cpp:153]: @samdow -- how do we handle randn_like, rand_like right now?

File: functorch/csrc/BatchRulesRandomness.cpp [hunk 3]
```diff
@@ -149,6 +149,33 @@ Tensor unary_pointwise_random_batch_rule(const Tensor& tensor, ExtraArgs... extr
...
+  RandomnessType randomness = maybe_layer->randomness();
+  check_randomness(randomness);
+
+  Tensor tensor_value;
+  optional<int64_t> tensor_bdim;
+  std::tie(tensor_value, tensor_bdim) = unwrapTensorAtLevel(self, cur_level);
+  tensor_value = moveBatchDimToFront(tensor_value, tensor_bdim);
+
+  if (randomness == RandomnessType::Same && tensor_bdim) {
+    tensor_value = tensor_value[0];
```
Comment (zou3519) [inline functorch/csrc/BatchRulesRandomness.cpp:167]: edge case, but what if tensor_vlaue has no dimensions? `tensor_value = torch.ones({})`

File: functorch/csrc/BatchRulesRandomness.cpp [hunk 4]
```diff
@@ -149,6 +149,33 @@ Tensor unary_pointwise_random_batch_rule(const Tensor& tensor, ExtraArgs... extr
...
+  } else if (randomness == RandomnessType::Different && !tensor_bdim) {
+    auto shape = tensor_value.sizes();
+    VmapDimVector shapeVec(shape.begin(), shape.end());
+    shapeVec.insert(shapeVec.begin(), maybe_layer->batchSize());
```
Comment (zou3519) [inline functorch/csrc/BatchRulesRandomness.cpp:171]: Nit: this insertion is O(n), we should insert first, then copy over the rest of the list

Comment (zou3519) [reply functorch/csrc/BatchRulesRandomness.cpp:171]: Also, shapeVec isn't being used?

File: test/test_vmap.py [hunk 5]
```diff
@@ -3581,7 +3566,7 @@ def test_feature_dropout(self, device, randomness):
         ]
 
         for op in supported_ops:
-            passed = torch.ones([4, 3, 3, 3, 3])
+            passed = torch.ones([4, 3, 3, 3, 3], device=device)
```
Comment (zou3519) [inline test/test_vmap.py:3569]: We should really test batch_dim != 0 as well as batch_dim == 0 (IIRC all the existing tests test batch_dim == 0)?

Comment (zou3519) [reply test/test_vmap.py:3569]: While I'm on the "things to improve testing train": we should have a one-off manual test with just a single op that tests multiple vmaps and different randomness behavior to check that the mode mechanism works with multiple levels.

That is, we should check `vmap(vmap(f, randomness=different), randomness=same)` in all 9 combinations of randomness = {same, different, error} x {same, different, error}.

File: functorch/csrc/VmapModeRegistrations.cpp [hunk 6]
```diff
@@ -5,39 +5,15 @@
...
 
 
 namespace at {
 namespace functorch {
 
-void unsupportedRandomOp(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
-  TORCH_CHECK(false, "vmap: We do not yet support calling random operations inside of vmap. ",
-              "Please perform random operations outside of vmap as a workaround");
-}
-
```
Comment (zou3519) [inline functorch/csrc/VmapModeRegistrations.cpp:23]: (no action required) I can see some future where when someone adds a new randomness function to native_functions.yaml, we automatically generate an "unsupported" error message. But we can bring this back when that happens

**Review by zou3519 — COMMENTED** [review]
some questions & testing suggestions

File: functorch/csrc/BatchRulesRandomness.cpp [hunk 7]
```diff
@@ -149,6 +149,33 @@ Tensor unary_pointwise_random_batch_rule(const Tensor& tensor, ExtraArgs... extr
...
+    tensor_value = tensor_value.unsqueeze(0).expand(batch_size);
```
Comment (zou3519) [inline functorch/csrc/BatchRulesRandomness.cpp:172]: Is this assuming `tensor_value` has 0 dimensions?

File: functorch/csrc/BatchRulesRandomness.cpp (hunk 2 above)
Comment (samdow) [reply functorch/csrc/BatchRulesRandomness.cpp:153]: Good point--didn't see those in the original randomness error message but I'll double check them and probably end up adding them here

File: functorch/csrc/BatchRulesRandomness.cpp (hunk 3 above)
Comment (samdow) [reply functorch/csrc/BatchRulesRandomness.cpp:167]: Would we be able to vmap over a tensor with no dimensions?

File: functorch/csrc/BatchRulesRandomness.cpp (hunk 7 above)
Comment (samdow) [reply functorch/csrc/BatchRulesRandomness.cpp:172]: Good call--I think this was supposed to use shapeVec and now I'm a little confused how it works (probably another sign to reassess the randomness testing)

**Comment by samdow** [comment]
Going to put in a fix for dropout running on an unbatched input. This is going to stack on top of that so this will be paused (but will fix shapeVec usage + add other like functions while the dropout fix is testing + landing)

File: functorch/csrc/BatchRulesRandomness.cpp (hunk 3 above)
Comment (zou3519) [reply functorch/csrc/BatchRulesRandomness.cpp:167]: Aha, good point. You're right -- this won't happen and tensor_value will have at least one dimension
